src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsstafflookupservice/service/IndexingService.kt

```diff
@@ -1,21 +1,57 @@
 package uk.gov.justice.digital.hmpps.hmppsstafflookupservice.service
 
+import com.microsoft.applicationinsights.TelemetryClient
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
 import org.springframework.stereotype.Service
+import org.springframework.util.StopWatch
 import uk.gov.justice.digital.hmpps.hmppsstafflookupservice.client.MicrosoftGraphClient
+import uk.gov.justice.digital.hmpps.hmppsstafflookupservice.service.telemetry.TelemetryEventType
 
 @Service
 class IndexingService(
+  val statusStore: StatusStore,
+  val telemetryClient: TelemetryClient,
   val microsoftGraphClient: MicrosoftGraphClient,
   val databaseWriteService: DatabaseWriteService
 ) {
-  suspend fun indexAll(): String {
-    var graphResponse = microsoftGraphClient.getUsersPage(null)
-    databaseWriteService.writeData(graphResponse.value)
-    while (graphResponse.nextLink != null) {
-      val skipToken = graphResponse.nextLink!!.substringAfter("skiptoken=")
-      graphResponse = microsoftGraphClient.getUsersPage(skipToken)
+  fun indexAll() {
+    executeAsynchronously {
+      var graphResponse = microsoftGraphClient.getUsersPage(null)
       databaseWriteService.writeData(graphResponse.value)
+      while (graphResponse.nextLink != null) {
+        val skipToken = graphResponse.nextLink!!.substringAfter("skiptoken=")
+        graphResponse = microsoftGraphClient.getUsersPage(skipToken)
+        databaseWriteService.writeData(graphResponse.value)
+      }
+    }
+    return
+  }
+
+  private fun executeAsynchronously(indexingBlock: suspend () -> Unit) {
+    statusStore.checkAndSetInProgress()
+    CoroutineScope(Dispatchers.IO).launch {
```

Comment:
The choice of `IO` dispatcher was a bit arbitrary - but it sounded like the most flexible